packages/firestore/src/remote/remote_store.ts

```diff
@@ -136,82 +152,65 @@ export class RemoteStore implements TargetMetadataProvider {
-      'WatchStream and WriteStream should both be null or non-null'
-    );
-    return this.watchStream != null;
+    await this.startStreamsIfNecessary();
   }
 
   /** Re-enables the network. Idempotent. */
-  enableNetwork(): Promise<void> {
-    if (this.isNetworkEnabled()) {
-      return Promise.resolve();
-    }
-
-    // Create new streams (but note they're not started yet).
-    this.watchStream = this.datastore.newPersistentWatchStream();
-    this.writeStream = this.datastore.newPersistentWriteStream();
-
-    // Load any saved stream token from persistent storage
-    return this.localStore.getLastStreamToken().then(token => {
-      this.writeStream.lastStreamToken = token;
+  async enableNetwork(): Promise<void> {
+    this.networkEnabled = true;
+    await this.startStreamsIfNecessary();
+  }
 
-      if (this.shouldStartWatchStream()) {
-        this.startWatchStream();
-      } else {
-        this.onlineStateTracker.set(OnlineState.Unknown);
-      }
+  private async startStreamsIfNecessary(): Promise<void> {
+    if (this.shouldStartWatchStream()) {
+      this.startWatchStream();
+    } else {
+      this.onlineStateTracker.set(OnlineState.Unknown);
+    }
 
-      return this.fillWritePipeline(); // This may start the writeStream.
-    });
+    // This will start the write stream if necessary.
+    await this.fillWritePipeline();
   }
 
   /**
    * Temporarily disables the network. The network can be re-enabled using
    * enableNetwork().
    */
   async disableNetwork(): Promise<void> {
-    this.disableNetworkInternal();
+    this.networkEnabled = false;
+    this.stopStreams();
+
     // Set the OnlineState to Offline so get()s return from cache, etc.
     this.onlineStateTracker.set(OnlineState.Offline);
   }
 
-  /**
-   * Disables the network, if it is currently enabled.
-   */
-  private disableNetworkInternal(): void {
-    if (this.isNetworkEnabled()) {
-      // NOTE: We're guaranteed not to get any further events from these streams (not even a close
-      // event).
-      this.watchStream.stop();
-      this.writeStream.stop();
-
-      this.cleanUpWatchStreamState();
-
+  private stopStreams(): void {
+    if (this.writePipeline.length > 0) {
       log.debug(
         LOG_TAG,
         'Stopping write stream with ' +
           this.writePipeline.length +
           ' pending writes'
       );
-      // TODO(mikelehen): We only actually need to clear the write pipeline if
-      // this is being called as part of handleUserChange(). Consider reworking.
-      this.writePipeline = [];
-
-      this.writeStream = null;
-      this.watchStream = null;
     }
+
+    // Temporarily set networkEnabled to false so that our close handling
+    // doesn't try to restart the streams, since we are explicitly closing them.
+    const networkEnabled = this.networkEnabled;
+    this.networkEnabled = false;
+    this.writeStream.stop();
+    this.watchStream.stop();
```

Comment:
Do we still need to restart the stream if we did not receive an error? It seems like that might be enough to decide whether we should restart or not.

Reply:
So that is a potential option, though I would be nervous about relying on that...  And right now I actually have an assert that explicitly makes sure that shouldStartWatchStream() returns false if we did not receive an error (see my discussion with Gil about that), so if I remove the "this.networkEnabled = false;" line, we hit that assert.